### PR TITLE
feat(core): Allow transferring workflows from any project to any team project

### PR DESCRIPTION
**Pull request by despairblue**
## Summary

This allows transferring workflows from any personal or team project to any other team project.

The user needs to have the permissions/scopes of `workflow:move` for the workflow and `workflow:create` for the destination project.

Workflows cannot be moved into their current home project.
Workflows cannot be moved into any personal project.
Workflows can only be moved out of personal and team projects. (For now there are only personal and team projects, but there are plans for other kinds of projects.)

## Related tickets and issues

https://www.notion.so/n8n/Move-resources-workflow-credentials-between-projects-e737365c6ff446ce9e396df03e7a1255
https://linear.app/n8n/issue/DOC-886/feature-transfer-workflows-and-credentials-from-project-to-project
https://linear.app/n8n/issue/PAY-1428/move-resources-between-projects

## Review / Merge checklist

- [x] PR title and summary are descriptive. **Remember, the title automatically goes into the changelog. Use `(no-changelog)` otherwise.** ([conventions](https://github.com/n8n-io/n8n/blob/master/.github/pull_request_title_conventions.md))
- [x] Tests included.
  > A bug is not considered fixed, unless a test is added to prevent it from happening again.
  > A feature is not complete without tests.

### CODE REVIEW

File: packages/cli/src/workflows/workflow.service.ee.ts [hunk 1]
```diff
@@ -235,4 +241,98 @@ export class EnterpriseWorkflowService {
...
+
+	async transferOne(user: User, workflowId: string, toProjectId: string) {
+		// 1. get workflow
+		const workflow = await this.sharedWorkflowRepository.findWorkflowForUser(workflowId, user, [
+			'workflow:delete',
+		]);
+		NotFoundError.isDefinedAndNotNull(
+			workflow,
+			`Could not find workflow with the id "${workflowId}". Make sure you have the permission to delete it.`,
+		);
```
Comment (despairblue) [inline packages/cli/src/workflows/workflow.service.ee.ts:253]: I find type guards easier to read than conditional logic and creating an error. 

But I don't feel strongly about it, so it's up for debate.

File: packages/workflow/src/errors/abstract/execution-base.error.ts [hunk 2]
```diff
@@ -37,8 +37,7 @@ export abstract class ExecutionBaseError extends ApplicationError {
 		if (errorResponse) this.errorResponse = errorResponse;
 	}
 
-	// eslint-disable-next-line @typescript-eslint/no-explicit-any
-	toJSON?(): any {
```
Comment (despairblue) [inline packages/workflow/src/errors/abstract/execution-base.error.ts:41]: Not sure why we have to disable the return type inference here.

**Comment by StarfallProjects** [comment]
@despairblue this one probably needs docs - can you create a ticket and link relevant info/dev tickets etc.

File: packages/cli/src/workflows/workflows.controller.ts [hunk 3]
```diff
@@ -460,4 +470,16 @@ export class WorkflowsController {
 			workflow,
 		});
 	}
+
+	@Put('/:workflowId/transfer')
+	@Licensed('feat:advancedPermissions')
```
Comment (despairblue) [inline packages/cli/src/workflows/workflows.controller.ts:475]: No sure about this, we've put lots of other endpoints behind the admin feature flag.

**Comment by despairblue** [comment]
> @despairblue this one probably needs docs - can you create a ticket and link relevant info/dev tickets etc.

For now this is only the internal API side. But I can create a doc ticket and link the notion page with the mock-ups, sure thing!

File: packages/cli/src/workflows/workflows.controller.ts (hunk 3 above)
Comment (despairblue) [reply packages/cli/src/workflows/workflows.controller.ts]: It was decided that this endpoint is not behind a license flag.

File: packages/cli/src/workflows/workflows.controller.ts [hunk 4]
```diff
@@ -460,4 +461,16 @@ export class WorkflowsController {
 			workflow,
 		});
 	}
+
+	@Put('/:workflowId/transfer')
+	@ProjectScope('workflow:move')
```
Comment (krynble) [inline packages/cli/src/workflows/workflows.controller.ts:466]: Not sure how this validation would work - would it check the user's personal project or the target project scope?

File: packages/cli/test/integration/workflows/workflows.controller.ee.test.ts [hunk 5]
```diff
@@ -1236,3 +1242,308 @@ describe('PATCH /workflows/:workflowId - activate workflow', () => {
...
+		const workflow = await createWorkflow({}, member);
+
+		await testServer
+			.authAgentFor(member)
+			.put(`/workflows/${workflow.id}/transfer`)
+			.send({ destinationProjectId: destinationProject.id })
+			.expect(404);
+	});
+
+	test('project members cannot transfer workflows', async () => {
```
Comment (krynble) [inline packages/cli/test/integration/workflows/workflows.controller.ee.test.ts:1295]:
```suggestion
	test('user should not be able to transfer workflows to projects they are not part of', async () => {
```

File: packages/cli/test/integration/workflows/workflows.controller.ee.test.ts [hunk 6]
```diff
@@ -1236,3 +1242,308 @@ describe('PATCH /workflows/:workflowId - activate workflow', () => {
...
+		// ACT & ASSERT
+		//
+		await testServer
+			.authAgentFor(member)
+			.put(`/workflows/${workflow.id}/transfer`)
+			.send({ destinationProjectId: destinationProject.id })
+			.expect(403);
+	});
+
+	test('can transfer from personal to team project', async () => {
```
Comment (krynble) [inline packages/cli/test/integration/workflows/workflows.controller.ee.test.ts:1315]:
```suggestion
	test('can transfer from personal to team project and remove existing sharing', async () => {
```

Comment (despairblue) [reply packages/cli/test/integration/workflows/workflows.controller.ee.test.ts:1315]: I think this is implicit behavior. A move will always sever all sharings, it's not optional. 

But I can split the tests to make one called: `moving from a personal project to a team project removes all existing sharings`.

File: packages/cli/test/integration/workflows/workflows.controller.ee.test.ts (hunk 5 above)
Comment (despairblue) [reply packages/cli/test/integration/workflows/workflows.controller.ee.test.ts:1295]: 😬 I'll improve the name. This test tests that a `project:editor` cannot transfer workflows out of projects. That check happens before we check that the user has the `workflow:create` scope in the destination project so I didn't actually link the user to it. I'll amend that.

**Review by krynble — DISMISSED** [review]

**Review by krynble — APPROVED** [review]

**Comment by janober** [comment]
Got released with [`n8n@1.45.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.45.0)

**Pull request merged**